## Numpy and pandas


In [1]:
import numpy as np
import pandas as pd

### Question 1 (20 points) Polynomial function

Consider the polynomial expression

$p(x)=a_0+a_1x+a_2x^2+⋯a_Nx^N=\sum_{n=0}^{N}a_nx^n$

**Question 1.1 (5 points)** Implement a simple function p(x, alphas) to compute the value given a point x and a list of alphas, using **enumerate()** in the for loop:

Example input: p(2, (1,2,3))

Example output: 17

In [ ]:

def p(x,alphas):
    alphas=tuple(int(i) for i in alphas)
    y = 0
    for i, a in enumerate(alphas):
        y = y + a*int(x)**i 
    return "The result is: {}".format(y)
p(2, (1,2,3))

**Explanation Q1_1:** the function creates a tuple which contain the values of alphas. Then by using "enumerates" it is able to index the alphas and use those index to power the *x*.

**Question 1.2 (5 points)**  Now write a new function that does the same job, but uses numpy function np.poly1d to compute the value: 


In [5]:
def p(x,alphas):
    alphas=[int(i) for i in alphas][::-1]
    x=int(x)
    fun=np.poly1d(alphas)
    return "The result is: {}".format(fun(x))
p(2, (1,2,3))

'The result is: 17'

**Explanation Q1_2:** the function creates an inverted list of constants "alphas" and then apply to the numpy function *poly1d*.


**Question 1.3 (10 points)** Now write a new function that does the same job, but uses NumPy arrays and array operations for its computations (not np.poly1d), rather than any form of Python loop. 

In [6]:
def p(x,alphas):
    alphas=list(int(i) for i in alphas)
    a=np.array([alphas])
    x_n=np.repeat(x,len(alphas))
    x_n=x_n.astype("int32")
    power=np.arange(len(alphas))
    res=np.sum((x_n**power)*a)
    return "The result is: {}".format(res)
    
p(2, (1,2,3))

'The result is: 17'

**Explanation Q1_3:** the function generates a list of alphas. Then it creates a vector *x_n* which contains the *x* value repeated as many time as the length of alphas. Finally, it prints the sum of the *x* power to *n* multiply by the alphas.

### Question 2 (20 points) Numpy is fast!

Suppose we need to compute the cumulative sum of $\sum_{i=0}^n \alpha^i$ for given $\alpha$ and $n$. 

For example, when $\alpha=0.5$ and $n=10$, the cumulative sum of $\sum_{i=0}^{10} 0.5^i$ returns `[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375]`


As a courtesy, I implement the following function `cum_sum` that can be used to generate a list of cumulative sum when iterating through a range generator.

In [123]:
def cum_sum(alpha, n):
    current = 1.0
    sum = current
    for i in range(n):
        current = current * alpha
        sum = sum + current
    return sum 

cumsum = []
for i in range(11):
    cumsum.append(cum_sum(0.5, i))
    
print(cumsum)

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375]


We can calculate how much time does it spend to run this code using `time` module as below:

In [124]:
import time

begin = time.time()
n_samples = 10000

cumsum = []
for i in range(n_samples):
    cumsum.append(cum_sum(0.5, i))
print(cumsum)
    
end = time.time()

time0 = end-begin
print("Time took to run: {} seconds.".format(time0))

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375, 1.99951171875, 1.999755859375, 1.9998779296875, 1.99993896484375, 1.999969482421875, 1.9999847412109375, 1.9999923706054688, 1.9999961853027344, 1.9999980926513672, 1.9999990463256836, 1.9999995231628418, 1.999999761581421, 1.9999998807907104, 1.9999999403953552, 1.9999999701976776, 1.9999999850988388, 1.9999999925494194, 1.9999999962747097, 1.9999999981373549, 1.9999999990686774, 1.9999999995343387, 1.9999999997671694, 1.9999999998835847, 1.9999999999417923, 1.9999999999708962, 1.999999999985448, 1.999999999992724, 1.999999999996362, 1.999999999998181, 1.9999999999990905, 1.9999999999995453, 1.9999999999997726, 1.9999999999998863, 1.9999999999999432, 1.9999999999999716, 1.9999999999999858, 1.999999999999993, 1.9999999999999964, 1.9999999999999982, 1.9999999999999991, 1.9999999999999996, 1.9999999999999998, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 

It takes about 3.6 seconds on my machine to run the code with 10,000 samples. Note that this time may vary depending on the memory and CPU of your machine. 

**Question 2.1** (15 points) Now implement a list comprehension for the same purpose and estimate how much time does it take to generate a list of cumulative sum for 10,000 samples

Hint: you can use the method accumulate in the module itertools. Check the documentation of itertools at [here](https://docs.python.org/3/library/itertools.html#itertools.accumulate)

In [129]:
# Question 2.1
import time
import itertools as it

def my_fun(alpha,n):
    l=[alpha**i for i in range(n)]
    l=list(it.accumulate(l))
    print(l)
    return l
my_fun(0.5,10)

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875]


[1.0,
 1.5,
 1.75,
 1.875,
 1.9375,
 1.96875,
 1.984375,
 1.9921875,
 1.99609375,
 1.998046875]

In [130]:
begin = time.time()
n_samples = 10000

# write your code here
my_fun(0.5,n=n_samples)

end = time.time()

time1 = end-begin

print("Time took to run: {} seconds.".format(time1))

[1.0, 1.5, 1.75, 1.875, 1.9375, 1.96875, 1.984375, 1.9921875, 1.99609375, 1.998046875, 1.9990234375, 1.99951171875, 1.999755859375, 1.9998779296875, 1.99993896484375, 1.999969482421875, 1.9999847412109375, 1.9999923706054688, 1.9999961853027344, 1.9999980926513672, 1.9999990463256836, 1.9999995231628418, 1.999999761581421, 1.9999998807907104, 1.9999999403953552, 1.9999999701976776, 1.9999999850988388, 1.9999999925494194, 1.9999999962747097, 1.9999999981373549, 1.9999999990686774, 1.9999999995343387, 1.9999999997671694, 1.9999999998835847, 1.9999999999417923, 1.9999999999708962, 1.999999999985448, 1.999999999992724, 1.999999999996362, 1.999999999998181, 1.9999999999990905, 1.9999999999995453, 1.9999999999997726, 1.9999999999998863, 1.9999999999999432, 1.9999999999999716, 1.9999999999999858, 1.999999999999993, 1.9999999999999964, 1.9999999999999982, 1.9999999999999991, 1.9999999999999996, 1.9999999999999998, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 

In [131]:
time0/time1

65.12884656201243

**Question 2.2** (15 points) Now implement using numpy for the same purpose and estimate how much time does it take to generate a list of cumulative sum for 10,000 samples (in order to receive full score, your program must be at least 1000 times faster than the for loop)


In [1]:
import numpy as np

In [117]:
# Question 2.2

begin = time.time()
n_samples = 1000000
alpha = 0.5
# write your code here

n_samples = 1000000
alpha = 0.5

u = np.empty(n_samples)
u[0] = 1
u[1:] = alpha

cumsum = np.cumprod(u)

cumsum = np.cumsum(cumsum)


print(cumsum)
end = time.time()

time2 = end-begin

print("Time took to run: {} seconds.".format(time2))

[1.   1.5  1.75 ... 2.   2.   2.  ]
Time took to run: 0.02748847007751465 seconds.


In [118]:
# Question 2.2

begin = time.time()
n_samples = 10000000
alpha = 0.5
# write your code here

def my_fun(alpha,n):
    l=[alpha**i for i in range(n+1)]
    l=np.cumsum(l)
    print(l)
    return l
my_fun(0.5,n_samples)


end = time.time()

time2 = end-begin

print("Time took to run: {} seconds.".format(time2))

[1.   1.5  1.75 ... 2.   2.   2.  ]
Time took to run: 6.611839532852173 seconds.


In [119]:
time0/time2

NameError: name 'time0' is not defined

### Question 3 (30 points) Monte Carlo 

Monte Carlo is a city in Monacco where the famous Monte Carlo casino is located.

![casino](https://www.casinonewsdaily.com/wp-content/uploads/2014/12/The-Casino-de-Monte-Carlo.jpg)

In light of this, Monte Carlo methods (or Monte Carlo experiments) are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. Their essential idea is using randomness to solve problems that might be deterministic in principle. They are often used in physical and mathematical problems and are most useful when it is difficult or impossible to use other approaches. Monte Carlo methods are mainly used in three problem classes: optimization, numerical integration, and generating draws from a probability distribution.

**Estimate the Pi**

In order to estimate the $\pi$, the idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. Imagine a circle inside the same domain with same diameter and inscribed into the square. We can generate a large number of uniformly distributed random points and plot them on the graph. These points can be in any position within the square i.e. between (0,0) and (1,1). We keep track of the total number of points, and the number of points that are inside the circle. If we divide the number of points within the circle, $N_{inner}$ by the total number of points, $N_{total}$, we should get a value that is an approximation of the ratio of the areas we calculated above, $\pi/4$.

Write a function `approximate_pi` with argument `number_simulations` to approximate the Pi value using Monte Carlo simulations. You may consider to use `numpy.random` to make random draws. 

Give a rough estimate about how many random draws you may need to achieve accuray of 99.9% (by comparing with numpy.pi).

In [116]:
import numpy as np
np.pi

3.141592653589793

In [88]:
# Question 3
import numpy as np

def MonteCarlo_pi(n_samples):
    count = 0
#Generate a bunch of values of x and y between -1 and 1, then assess their combined radius on an xy plane
    for i in range(n_samples):
        x = 2*(np.random.random()-0.5)
        y = 2*(np.random.random()-0.5)
        r = np.sqrt(x**2+y**2)
        if r <= 1:
            count +=1
        else:
            continue
    global pi, distance        
    pi = 4*(count / n_samples)
    distance = abs(100*((pi-np.pi)/np.pi))
    
    return "The value of pi found is {}, it is distant from the real value by {} %".format(pi, distance)
    
MonteCarlo_pi(10000)

'The value of pi found is 3.1508, it is distant from the real value by 0.29307893878876307 %'

In [89]:
for i in range(1,100000000,100):
    MonteCarlo_pi(i)
    if distance <=0.01:
        print("I have found pi with the accuracy of 99.9% roughly at the {} iterations".format(i))
        print(pi)
        break
    else:
        continue

I have found pi with the accuracy of 99.9% roughly at the 8501 iterations
3.1417480296435714


### Question 4 (30 points) California housing

We will explore the famous Califronia housing dataset. The original database is available from StatLib http://lib.stat.cmu.edu/datasets/


The data contains 20,640 observations on 9 variables.
This dataset contains the average house value as target variable
and the following input variables (features): records
the following for each tract in California: Median house price, median
house age, average number of rooms per house, average number of bedrooms,
average number of occupants, total number of houses, median income
(in thousands of dollars), latitude and longitude.

You can download the data from the Internet as:

```python
from sklearn.datasets.california_housing import fetch_california_housing

cal_housing = fetch_california_housing()

```

The dataset has the following format:
```
    dataset : dict-like object with the following attributes:
    dataset.data : ndarray, shape [20640, 8]
        Each row corresponding to the 8 feature values in order.
    dataset.target : numpy array of shape (20640,)
        Each value corresponds to the average house value in units of 100,000.
    dataset.feature_names : array of length 8
        Array of ordered feature names used in the dataset.
    dataset.DESCR : string
        Description of the California housing dataset.
```

In [61]:
from sklearn.datasets.california_housing import fetch_california_housing
cal_housing = fetch_california_housing()

In [49]:
wrong_california_housing = pd.DataFrame(data= np.c_[cal_housing['data'], cal_housing['target']], 
                                        ## concatenate the data with the target
                     columns= cal_housing['feature_names'] + ['target'])
                                        ## assign the col names
wrong_california_housing.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

**Question 4.1** (5 points) Create a dataframe of `california_housing` from the fetched dataset using the data, target and feature_names, shows how many observations/features

In [55]:
import pandas as pd
california_housing=pd.read_csv("housing.csv")
california_housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [77]:
print("Observations per features:\n",pd.DataFrame.count(california_housing))

Observations per features:
 longitude             20640
latitude              20640
housing_median_age    20640
total_rooms           20640
total_bedrooms        20433
population            20640
households            20640
median_income         20640
median_house_value    20640
ocean_proximity       20640
dtype: int64
longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64


In [80]:
print("Check how many missing values there are per features: \n",california_housing.isnull().sum())

Check how many missing values there are per features: 
 longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64


In [51]:
print("The data frame contain {} rows and {} columns".format(california_housing.shape[0],california_housing.shape[1]))

The data frame contain 20640 rows and 10 columns


**Question 4.2** (5 points) Show the descripive statistics of features

In [53]:
california_housing.describe().T

count           mean            std         min  \
longitude           20640.0    -119.569704       2.003532   -124.3500   
latitude            20640.0      35.631861       2.135952     32.5400   
housing_median_age  20640.0      28.639486      12.585558      1.0000   
total_rooms         20640.0    2635.763081    2181.615252      2.0000   
total_bedrooms      20433.0     537.870553     421.385070      1.0000   
population          20640.0    1425.476744    1132.462122      3.0000   
households          20640.0     499.539680     382.329753      1.0000   
median_income       20640.0       3.870671       1.899822      0.4999   
median_house_value  20640.0  206855.816909  115395.615874  14999.0000   

                            25%          50%           75%          max  
longitude             -121.8000    -118.4900    -118.01000    -114.3100  
latitude                33.9300      34.2600      37.71000      41.9500  
housing_median_age      18.0000      29.0000      37.00000      52.0000  
total_rooms           1447.7500    2127.0000    3148.00000   39320.0000  
total_bedrooms         296.0000     435.0000     647.00000    6445.0000  
population             787.0000    1166.0000    1725.00000   35682.0000  
households             280.0000     409.0000     605.00000    6082.0000  
median_income            2.5634       3.5348       4.74325      15.0001  
median_house_value  119600.0000  179700.0000  264725.00000  500001.0000

**Question 4.3** (10 points) Compare the areas with houses having more than 25 years old with areas with houses having less than 10 years old, what do you find about their total rooms, total bedrooms, median household incomes?

In [88]:
data=california_housing.loc[:,["housing_median_age","total_rooms","total_bedrooms","median_income"]]
data=data.iloc[:,:][data.housing_median_age>25]
summary_h25=data.describe().T
summary_h25

In [99]:
data=california_housing.loc[:,["housing_median_age","total_rooms","total_bedrooms","median_income"]]
data=data.iloc[:,:][data.housing_median_age<10]
summary_l10=data.describe().T
summary_l10

count         mean          std    min       25%  \
housing_median_age  1305.0     6.105747     2.039376  1.000     5.000   
total_rooms         1305.0  4957.250575  4748.846471  2.000  2029.000   
total_bedrooms      1293.0   911.307038   811.887666  2.000   389.000   
median_income       1305.0     4.656035     1.946202  0.536     3.375   

                          50%        75%         max  
housing_median_age     6.0000     8.0000      9.0000  
total_rooms         3484.0000  6190.0000  39320.0000  
total_bedrooms       691.0000  1155.0000   6210.0000  
median_income          4.2917     5.5023     15.0001

In [101]:
summary_h25-summary_l10

count         mean          std      min       25%  \
housing_median_age  10705.0    31.372771     5.754202  25.0000   27.0000   
total_rooms         10705.0 -2907.662981 -3585.522847   6.0000 -727.0000   
total_bedrooms      10585.0  -475.768563  -554.686623  -1.0000 -116.0000   
median_income       10705.0    -0.947114    -0.063004  -0.0361   -0.9179   

                          50%          75%      max  
housing_median_age    30.0000    35.000000     43.0  
total_rooms        -1638.0000 -3618.000000 -21582.0  
total_bedrooms      -304.0000  -613.000000  -3096.0  
median_income         -0.9167    -1.018225      0.0

**Findings:**  
* on average the older clusters of houses have got less rooms than the new houses by aproximatly 2908;
* the total number of bedrooms have got the same logic as before, holder cluster have lower number of bedrooms than the new, by 476;
* the households' income in older houses is lower than the new, a possible explanation is that high income profiles purchase new real estate.

**Question 4.4** (10 points) What is the difference between average median house price for households with top 20% of median houshold income and bottom 20% of median household income?

In [111]:
logic=np.percentile(california_housing['median_income'],80)
mean_high=np.mean(california_housing[california_housing.median_income>logic]["median_house_value"])
mean_high

334990.99781976745

In [115]:
logic=np.percentile(california_housing['median_income'],20)
mean_low=np.mean(california_housing[california_housing.median_income<logic]["median_house_value"])
mean_low

118385.48642753271

In [116]:
print("The difference between average median_house_value between top and low 20 is: " ,mean_high-mean_low)

The difference between average median_house_value between top and low 20 is:  216605.51139223474
